File: modules/navigator.js
/**
 * 页面导航模块
 * 
 * 负责在企业微信中导航：进入群聊、返回消息列表等
 */

var config = require("../config.js");
var utils = require("./utils.js");

var navigator = {

    /**
     * 回到企业微信消息列表首页
     * 通过反复按返回键直到到达消息列表
     */
    goToMessageList: function () {
        utils.log("正在返回消息列表...");
        var maxRetries = 10;
        for (var i = 0; i < maxRetries; i++) {
            // 检查是否已经在消息列表页面
            // 企业微信底部标签栏有"消息"按钮
            var msgTab = text("消息").findOne(1000);
            if (msgTab) {
                // 点击"消息"标签确保在消息列表
                utils.safeClick(msgTab);
                utils.debug("已到达消息列表页面");
                return true;
            }

            // 还没到，按返回键
            utils.goBack();
            utils.wait(500);
        }

        // 兜底：直接启动企业微信
        utils.log("无法返回消息列表，尝试重新启动企业微信");
        app.launchPackage(config.packageName);
        utils.wait(3000);

        var msgTab = text("消息").findOne(3000);
        if (msgTab) {
            utils.safeClick(msgTab);
            return true;
        }

        utils.log("返回消息列表失败！");
        return false;
    },

    /**
     * 从消息列表进入指定群聊
     * @param {string} groupName 群名称
     * @returns {boolean} 是否成功进入
     */
    enterGroup: function (groupName) {
        utils.log("正在进入群聊: " + groupName);

        // 先确保在消息列表
        if (!navigator.goToMessageList()) {
            return false;
        }

        utils.wait(500);

        // 在消息列表中查找群名
        // 先直接找，可能在当前屏幕上
        var group = navigator._findGroupInList(groupName);
        if (group) {
            utils.safeClick(group);
            utils.log("已点击进入群聊: " + groupName);
            // 等待群聊页面加载
            utils.waitSeconds(config.enterGroupWaitSeconds);
            return navigator._verifyInChatPage(groupName);
        }

        // 没找到，尝试搜索
        utils.log("列表中未找到群聊，尝试搜索: " + groupName);
        return navigator._searchAndEnterGroup(groupName);
    },

    /**
     * 在消息列表中查找群（包含向下滚动查找）
     */
    _findGroupInList: function (groupName) {
        // 先在当前屏幕找
        var group = text(groupName).findOne(1500);
        if (group) return group;

        // 向下滑动几次查找
        for (var i = 0; i < 5; i++) {
            utils.swipeDown();
            group = text(groupName).findOne(1000);
            if (group) return group;
        }

        return null;
    },

    /**
     * 通过搜索功能进入群聊
     */
    _searchAndEnterGroup: function (groupName) {
        // 点击搜索框/搜索图标
        // 企业微信顶部一般有搜索区域
        var searchBox = id("ici").findOne(2000)    // 搜索框常见 id
            || desc("搜索").findOne(2000)
            || text("搜索").findOne(2000);

        if (!searchBox) {
            utils.log("找不到搜索入口");
            return false;
        }

        utils.safeClick(searchBox);
        utils.wait(1000);

        // 在搜索框中输入群名
        var input = className("EditText").findOne(3000);
        if (!input) {
            utils.log("找不到搜索输入框");
            utils.goBack();
            return false;
        }

        input.setText(groupName);
        utils.wait(1500); // 等待搜索结果

        // 查找搜索结果中的群名
        var result = text(groupName).findOne(3000);
        if (!result) {
            utils.log("搜索结果中未找到群聊: " + groupName);
            utils.goBack();
            utils.goBack();
            return false;
        }

        utils.safeClick(result);
        utils.waitSeconds(config.enterGroupWaitSeconds);
        return navigator._verifyInChatPage(groupName);
    },

    /**
     * 验证是否成功进入了聊天页面
     */
    _verifyInChatPage: function (groupName) {
        // 聊天页面一般顶部有群名，底部有输入框
        var titleCheck = text(groupName).findOne(2000);
        var inputCheck = className("EditText").findOne(2000)
            || id("bhn").findOne(2000);  // 企业微信输入框 id

        if (titleCheck || inputCheck) {
            utils.debug("已确认进入群聊页面: " + groupName);
            return true;
        }

        utils.log("进入群聊验证失败: " + groupName);
        return false;
    },

    /**
     * 从群聊页面返回消息列表
     */
    exitGroup: function () {
        utils.debug("退出当前群聊...");
        utils.goBack();
        utils.wait(500);

        // 确认回到消息列表
        var msgTab = text("消息").findOne(2000);
        if (msgTab) {
            utils.debug("已返回消息列表");
            return true;
        }

        // 可能还在某个中间页面，再按一次返回
        utils.goBack();
        utils.wait(500);
        return !!text("消息").findOne(2000);
    },

    /**
     * 检查指定群是否有新消息未读
     * 在消息列表页面检查群名旁是否有未读标记
     * @param {string} groupName 群名称
     * @returns {boolean}
     */
    hasUnreadMessages: function (groupName) {
        var groupItem = text(groupName).findOne(1500);
        if (!groupItem) return false;

        // 检查群名所在行是否有未读角标
        // 企业微信的未读标记通常在群名同级或父容器中
        try {
            var parent = groupItem.parent();
            if (parent) {
                // 查找未读数字角标
                var badge = parent.findOne(className("TextView").textMatches("^\\d+$"));
                if (badge) {
                    utils.debug(groupName + " 有 " + badge.text() + " 条未读消息");
                    return true;
                }
                // 有些时候只显示红点，没有数字
                var redDot = parent.findOne(className("ImageView"));
                // 这里无法精确判断红点，简化处理
            }
        } catch (e) {
            utils.debug("检查未读消息失败: " + e);
        }

        return false;
    }
};

module.exports = navigator;

File: android/WeworkForwarder/app/src/main/java/com/wework/forwarder/Navigator.kt
package com.wework.forwarder

import android.content.Context
import android.content.Intent
import android.util.DisplayMetrics
import android.util.Log
import android.view.accessibility.AccessibilityNodeInfo

/**
 * 页面导航模块
 *
 * 负责在企业微信中导航：进入群聊、返回消息列表等
 */
object Navigator {
    private const val TAG = "Navigator"

    private fun log(msg: String) {
        Log.d(TAG, msg)
        Config.uiLog?.invoke(msg)
    }

    /**
     * 回到企业微信消息列表首页
     */
    fun goToMessageList(service: WeWorkAccessibilityService): Boolean {
        log("[导航] 返回消息列表...")
        for (i in 0 until 10) {
            val root = service.getRootNode()
            // 检查底部"消息"标签
            val msgTab = NodeFinder.findByText(root, "消息")
            if (msgTab != null) {
                NodeFinder.clickNode(service, msgTab)
                GestureHelper.delay()
                log("[导航] ✓ 已到达消息列表页面")
                return true
            }
            // 没到，按返回键
            service.pressBack()
            GestureHelper.delay(500)
        }

        // 兜底：直接启动企业微信
        log("[导航] 无法返回消息列表，重新启动企业微信")
        launchWeWork(service)
        GestureHelper.delayExact(3000)

        val root = service.getRootNode()
        val msgTab = NodeFinder.findByText(root, "消息")
        if (msgTab != null) {
            NodeFinder.clickNode(service, msgTab)
            return true
        }

        log("[导航] ✗ 返回消息列表失败")
        return false
    }

    /**
     * 从消息列表进入指定群聊
     */
    fun enterGroup(service: WeWorkAccessibilityService, metrics: DisplayMetrics, groupName: String): Boolean {
        log("[导航] 进入群聊: $groupName")

        if (!goToMessageList(service)) return false
        GestureHelper.delay(500)

        // 先在当前屏幕查找群名
        var group = findGroupInList(service, metrics, groupName)
        if (group != null) {
            NodeFinder.clickNode(service, group)
            log("[导航] ✓ 已点击进入群聊: $groupName")
            GestureHelper.delayExact(Config.enterGroupWaitSeconds * 1000L)
            return verifyInChatPage(service, groupName)
        }

        // 没找到，尝试搜索
        log("[导航] 列表中未找到，尝试搜索: $groupName")
        return searchAndEnterGroup(service, groupName)
    }

    /**
     * 在消息列表中查找群（包含向下滚动查找）
     */
    private fun findGroupInList(service: WeWorkAccessibilityService, metrics: DisplayMetrics, groupName: String): AccessibilityNodeInfo? {
        // 先在当前屏幕找
        var root = service.getRootNode()
        var group = NodeFinder.findByText(root, groupName)
        if (group != null) return group

        // 向下滑动几次查找
        for (i in 0 until 5) {
            GestureHelper.swipeDown(service, metrics)
            root = service.getRootNode()
            group = NodeFinder.findByText(root, groupName)
            if (group != null) return group
        }
        return null
    }

    /**
     * 通过搜索功能进入群聊
     */
    private fun searchAndEnterGroup(service: WeWorkAccessibilityService, groupName: String): Boolean {
        val root = service.getRootNode()

        // 点击搜索入口
        val searchBox = NodeFinder.findById(root, "ici")
            ?: NodeFinder.findByDesc(root, "搜索")
            ?: NodeFinder.findByText(root, "搜索")
        if (searchBox == null) {
            log("[导航] ✗ 找不到搜索入口")
            return false
        }
        NodeFinder.clickNode(service, searchBox)
        GestureHelper.delay(1000)

        // 在搜索框中输入群名
        val input = NodeFinder.waitForNode(service, 3000) { r ->
            NodeFinder.findByClassName(r, "android.widget.EditText")
        }
        if (input == null) {
            log("[导航] ✗ 找不到搜索输入框")
            service.pressBack()
            return false
        }
        NodeFinder.setText(input, groupName)
        GestureHelper.delayExact(1500)

        // 查找搜索结果中的群名
        val resultRoot = service.getRootNode()
        val result = NodeFinder.findByText(resultRoot, groupName)
            ?: NodeFinder.findByTextContains(resultRoot, groupName)
        if (result == null) {
            log("[导航] ✗ 搜索结果中未找到: $groupName")
            service.pressBack()
            service.pressBack()
            return false
        }
        NodeFinder.clickNode(service, result)
        GestureHelper.delayExact(Config.enterGroupWaitSeconds * 1000L)
        return verifyInChatPage(service, groupName)
    }

    /**
     * 验证是否成功进入了聊天页面
     */
    private fun verifyInChatPage(service: WeWorkAccessibilityService, groupName: String): Boolean {
        val root = service.getRootNode()
        val titleCheck = NodeFinder.findByText(root, groupName)
            ?: NodeFinder.findByTextContains(root, groupName)
        val inputCheck = NodeFinder.findByClassName(root, "android.widget.EditText")
        if (titleCheck != null || inputCheck != null) {
            log("[导航] ✓ 已确认进入群聊: $groupName")
            return true
        }
        log("[导航] ✗ 进入群聊验证失败: $groupName")
        return false
    }

    /**
     * 退出当前群聊
     */
    fun exitGroup(service: WeWorkAccessibilityService): Boolean {
        service.pressBack()
        GestureHelper.delay(500)
        val root = service.getRootNode()
        if (NodeFinder.findByText(root, "消息") != null) return true
        // 再按一次
        service.pressBack()
        GestureHelper.delay(500)
        return NodeFinder.findByText(service.getRootNode(), "消息") != null
    }

    /**
     * 确保企业微信在前台
     */
    fun ensureWeWorkForeground(service: WeWorkAccessibilityService): Boolean {
        if (service.isInWeWork()) return true
        log("[导航] 企业微信不在前台，正在启动...")
        launchWeWork(service)
        GestureHelper.delayExact(3000)
        return service.isInWeWork()
    }

    /**
     * v2.0: 在消息列表页扫描指定群的未读徽章数字
     *
     * 前提:已在消息列表页(由调用方 goToMessageList 保证)
     *
     * 返回:
     *   null  — 找不到群(列表上不存在,或滑了 5 屏仍未出现)
     *   0     — 群存在但无未读徽章
     *   K(>0) — 有 K 条未读;"99+" 折算为 99
     *
     * 实现:
     *   1. findGroupInList 风格的滚动查找(复用思路)
     *   2. 沿 parent 上行 4 级,定位群所在的"行容器"(childCount >= 3 的 ViewGroup)
     *   3. 行内 DFS 找匹配 ^\d+\+?$ 的 TextView 文本即为徽章数字
     *   4. 时间字段(如 "9:03"/"昨天")自然不匹配,无需额外排除
     */
    fun findUnreadCountForGroup(
        service: WeWorkAccessibilityService,
        metrics: DisplayMetrics,
        groupName: String
    ): Int? {
        var root = service.getRootNode() ?: return null
        var group = NodeFinder.findByText(root, groupName)
        var attempts = 0
        while (group == null && attempts < 5) {
            GestureHelper.swipeDown(service, metrics)
            root = service.getRootNode() ?: return null
            group = NodeFinder.findByText(root, groupName)
            attempts++
        }
        if (group == null) return null

        // 沿 parent 上行,找到包含群名 + 头像 + (徽章) + 时间 + 预览的行容器
        // 上行最多 4 级;一旦遇到 childCount >= 3 即认为是行容器
        var row: AccessibilityNodeInfo? = group.parent
        var levels = 0
        while (row != null && row.childCount < 3 && levels < 4) {
            row = row.parent
            levels++
        }
        val container = row ?: group

        // 在行内 DFS 提取所有文本,匹配纯数字徽章
        val texts = NodeFinder.getAllTexts(container)
        val badge = texts.firstNotNullOfOrNull { t ->
            val s = t.text.trim()
            when {
                s.matches(Regex("^\\d+$")) -> s.toIntOrNull()
                s.matches(Regex("^\\d+\\+$")) -> s.dropLast(1).toIntOrNull()
                else -> null
            }
        }
        return badge ?: 0
    }

    /**
     * 启动企业微信
     */
    private fun launchWeWork(context: Context) {
        try {
            val intent = context.packageManager.getLaunchIntentForPackage(Config.WEWORK_PACKAGE)
            intent?.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        } catch (e: Exception) {
            Log.e(TAG, "启动企业微信失败: ${e.message}")
        }
    }
}
